Chordality inspector added

Toptachamann opened the pull request:
Added chordality test. Main features are documented in Javadoc. In my implementation
I used following paper: 
* "CS 762: Graph-theoretic algorithms. Lecture notes of a graduate course. The University of Waterloo. Fall 1999, Winter 2002, Winter 2004". Visit this [link](http://www.cse.iitd.ac.in/~naveen/courses/CSL851/uwaterloo.pdf).

jkinable commented:
Make sure you copy&paste the project header into every file. Only change the year + your name. A header can be found in any class file

jsichi commented on jgrapht-core/src/test/java/org/jgrapht/alg/chordal/ChordalGraphInspectorTest.java:
```diff
@@ -0,0 +1,387 @@
+     * Tests proper creation of bucket list and removal of vertices from it
+     */
+    @Test
+    public void testBucketList1() {
```
Create a separate test class for each functional class.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/chordal/BucketList.java:
```diff
@@ -0,0 +1,281 @@
+ * @see ChordalGraphInspector
+ * @since 1.8
+ */
+public class BucketList<V> {
```
This class is currently just an internal data structure for the ChordalGraphInspector, so in its current form it shouldn't be public.

If you intend it to be reusable as a general LexicographicBFS, then it should be renamed and enhanced accordingly before being made public.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/chordal/BucketList.java:
```diff
@@ -0,0 +1,281 @@
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.chordal;
```
It might be a little to early to introduce a dedicated chordal package. For now I would include it in the cycle package (after all, a chordal graph is a graph in which all cycles of four or more vertices have a chord). Introducing a new (chordal) package can be done later when it becomes necessary.

jsichi replied:
makes sense

jkinable commented:
Thanks for your submission. Can you indicate which algorithm you used? You are referring to some CS lecture. Do you by any chance know which algorithm that is? Also, clearly state the runtime complexity. Is the algorithm by any chance based on this paper: http://epubs.siam.org/doi/abs/10.1137/0213035
Please be as precise as possible.
Also, in your search for a potential algorithm, did you encounter any references to faster/better algorithms? This would be good to know for a potential future extension.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/chordal/ChordalGraphInspector.java:
```diff
@@ -0,0 +1,191 @@
+     * perfect elimination order
+     */
+    private boolean isPerfectEliminationOrder(Graph<V, E> graph, List<V> vertexOrder, Map<V, Integer> map) {
+        Set<Pair<V, V>> testEdges = new HashSet<>();
```
The reason was not to check the presence of the same edge multiple times.
For example, if we have a 4-vertex graph G = (V, E): V = {1,2,3,4}, E = {{1,2},{1,3},{1,4},{2,3},{2,4}}; and vertexOrder = (1,2,3,4), then:
* first and second iterations give no edges;
* third and fourth iteration give us the same edge {1,2} two times.

Clearly, there are examples, where the same edge can be encountered more than 2 times. And it can be useful to have this buffer in cases, where we can't check the presence of an edge in constant time. But then I found out that FastLookupUndirectedSpecifics is used rather than UndirectedSpecifics (correct me if I'm wrong), which allows constant-time lookups.
So, now both edge addition to buffer and in-time presence check take constant time. Buffering edges increases space complexity to O(|E|). At the same time, it slows down the method in those cases, when we could almost immediately say, that vertices in vertexOrder aren't in the perfect elimination order by revealing a missing edge. 
So, yeah, there's no need to buffer those edges.

Toptachamann commented:
**Description of algorithm and citations**
I used an algorithm called lexicographical breadth-first search. It is described in [1] on pages 21-26. Complexity is linear, i.e. O(|V| + |E|). 
Description of the complexity of my implementation is placed in "complexity" section. Here I'm going to place a couple of useful citations I've found.
From [1]: 
"The first algorithm for finding a perfect elimination order is called Maximum Cardinality
Search or MCS, and was presented by Rose and Tarjan in 1975 [RT75]. The algorithm works
by repeatedly choosing the vertex for which the maximum number of neighbours has already
been chosen, breaking ties arbitrarily". Page 20.

"A second algorithm for finding a perfect elimination order is called lexicographic BFS or
lexBFS, and was presented by Rose, Tarjan, and Lueker in 1976 [RTL76]." Page 21. See [RTL76] in references. So, the algorithm I've implemented is based on paper [RTL76] and not on [2]. In CS lectures there are no references to the paper you've mentioned, namely [2].  


I've seen useful information in [3]:
"In the mid 1970s, Rose, Tarjan and Lueker [42] introduced a variant of BFS
called Lexicographic Breadth First Search (LBFS). Their application of LBFS
was to the recognition of chordal graphs. This algorithm is one of the classic
graph algorithms and, given the current interest in LBFS". Here [42] is a reference to [RTL76].

"Since determining whether a particular ordering is a perfect elimination ordering
can be accomplished in linear time, the algorithm has a straightforward
linear-time implementation. In a subsequent paper, Tarjan and Yannakakis [43]
extended this algorithm to be certifying by showing how to find, in linear time,
an induced cycle of size greater than three if the reverse of σ is not a perfect
elimination ordering". Here [43] is reference to [2].

In paper [4] I've found following paragraph:
"The adjacency-labeling concepts developed for the Lex-BFS algorithm have
proved to be central in the understanding of chordal graphs and triangulations.
Tarjan and Yannakakis later came up with the surprising result that for the
case of recognizing chordality, knowing the specific processed neighbors (i.e., labels)
is not necessary; one need only maintain and compare the cardinality of
processed neighbors [10]. This was a major breakthrough, resulting in a signifi-
cantly simplified implementation of Lex-BFS which has come to be known as the
MCS (maximum cardinality search) algorithm." Here [10] is reference to [2].

So, both lexicographical breadth-first search and maximum cardinality search give us elimination order on linear time. Determining whether a given order is a perfect elimination order takes linear time in both cases. So, both algorithms work in linear time. In the last citation, it is stated, that processing vertices with the maximum cardinality of already selected neighbours results in "significantly simplified implementation of Lex-BFS". I can't find a description of MCS (original article isn't freely accessible). The straightforward approach with priority queue will fail giving O((|V| + |E|)log|V|) time complexity and I can't come up with an idea how to fix it. At the same time, MCS gives us only an opportunity not to use auxiliary data structure. After getting rid of labels in LexBFS the difference between two algorithms is only a constant factor. At this point, it seems to me that running time won't differ significantly but I can't say for sure, cause I haven't read the original article.

Besides LexBFS and MCS I encountered only straightforward approach, that runs in O(|V|^4), what can't be the choice.  

**Complexity**
First of all, I've got rid of labels in my implementation. We only need them after selecting yet another vertex and putting it to order, while putting its neighbours in new buckets. Now the following approach is used: let v be the vertex that has been selected most recently. Every bucket that contains neighbours of v, we split into two. All vertices that are neightbours of v and were contained in the initial bucket we put in a new bucket and place it before initial bucket. If the initial bucket is now empty, we remove it from the data structure.

Complexity analyses of my implementation:
* Get elimination order:
  * for i =1 up to |V|: 
    * get vertex v from head of BucketList and append it to the end of the order -> O(1)
    * get neighbours N of vertex v that haven't been selected yet -> O(deg(v)). If we consider the whole algorithm, this step contributes O(|E|) time in aggregation, since we see each edge no more than 2 times during the whole algorithm.
    * Put vertices in N to new buckets. We create at most deg(v) new buckets (in practice this number is smaller) and put all O(deg(v)) vertices to this new buckets. Since we can find vertex's bucket in constant time, the whole process takes O(deg(v)) and contributes O(|E|) in aggregation to the total running time.
The total running time of getting elimination order if O(|V| + |E|).
* Check whether elimination order is perfect elimination order
  * for i =1 up to |V|: 
    * get i-th vertex in elimination order (let it be vertex v) -> O(1)
    * compute the predecessors in elimination order of vertex v -> O(deg(v))
    * compute predecessor with the maximum index in e.o. (let it be vertex predMax) -> O(deg(v))
    * check the presence of edge between predMax and all other predecessors of v -> O(deg(v))
The total running time of this procedure is O(|V| + |E|);
The total running time of the whole algorithm is therefore O(|V| + |E|)

References

[1] CS 762: Graph-theoretic algorithms. Lecture notes of a graduate course. University of Waterloo. Fall 1999, Winter 2002, Winter 2004.

[2] Robert E. Tarjan and Mihalis Yanakakis. Simple Linear-Time Algorithms to Test Chordality of Graphs, Test Acyclicity of Hypergraphs, and Selectively Reduce Acyclic Hypergraphs.

[3] Derek G. Corneil. Lexicographical Breadth First Search - A Survey

[4] Anne Berry, Jean R. S. Blair, and Pinar Heggernes. Maximum Cardinality Search for Computing Minimal Triangulations

[RT75] Donald J. Rose and R. Endre Tarjan. Algorithmic aspects of vertex elimination. In Seventh Annual ACM Symposium on Theory of Computing (Albuquerque, N. M., 1975), pages 245–254. Assoc. Comput. Mach., New York, 1975.

[RTL76] Donald J. Rose, R. Endre Tarjan, and George S. Lueker. Algorithmic aspects of vertex elimination on graphs. SIAM J. Comput., 5(2):266–283, 1976

jsichi commented on jgrapht-core/src/test/java/org/jgrapht/alg/cycle/BucketListTest.java:
```diff
@@ -0,0 +1,61 @@
+package org.jgrapht.alg.cycle;
```
Header is missing.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/BucketList.java:
```diff
@@ -0,0 +1,233 @@
+     */
+    BucketList(Collection<V> vertices) {
+        head = new Bucket(vertices);
+        bucketMap = new HashMap<>();
```
Optimize by giving the bucketMap an initial capacity

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
optimize capacity

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
"neighbors which have yet to be visited by..."

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,214 @@
+ * <a href="http://www.cse.iitd.ac.in/~naveen/courses/CSL851/uwaterloo.pdf">article</a>:
+ * <i>"CS 762: Graph-theoretic algorithms.
+ * Lecture notes of a graduate course. University of Waterloo. Fall 1999, Winter 2002, Winter 2004."</i>
+ * <p>
```
Since that doc covers a huge number of topics, including one of the more specific references from your previous comments would be good.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
If the lexicographical order will be computed lazily, the user won't be able to change a graph a little bit and then check whether the graph is still chordal. The only way to do it will be to create a new instance of the ChordalityInspector. On the other hand, when lex. order is recomputed after each call to isChordal(), single ChordalityInspector can be used for graph assigned to it. I can add a function like getChordality(), which will return last computed boolean chordal or compute it for the first time if it hasn't been done yet. 
What do you think?

jsichi replied:
Most of our existing algorithm classes are one-shot, without recomputation.  So recomputation should be explicit (e.g. rename this method to computeChordality, and use the name isChordal for the lazy version.

It's also possible to use a GraphListener to decide the recomputation automatically, although that's not a pattern we've followed much so far.

Toptachamann commented:
So, in [3] there was an action in the description of MCS:  "Choose an unnumbered vertex z of maximum weight". The actual way of doing this wasn't described, but it seems to me I understood how to do it. We can use that fact, that the cardinality of a vertex lies in [0, |V|-1] and sets of vertices can be easily enumerated with their cardinality. In the LexBFS this can't be accomplished even when labels are used implicitly due to the large number of possible vertex labels. This gives us the opportunity not to write auxiliary data structure.

I've already implemented MCS and tested the running time. It outperforms LexBFS by a factor of 1.3 - 1.4. I think this is due to the more complex updating stage after the addition of a new vertex to the order.

 I suggest following solution: MCS is used for testing chordality and ChordalityInspector receives common interface with CliqueMinimalSeparatorDecomposition after appropriate renaming. LexBFS in its current form is used to create LexBFSIterator and is moved to org.jgraph.traverse. LexBFSIterator will presumably extend BreadthFirstIterator.

jsichi commented:
In that case, sounds like we should close this pull request and start new ones.

1) MCS implementation.  Don't include the new interface in this one yet since there will probably be a bit of discussion around that.

2) LexicalBreadthFirstIterator.  I doubt you'll want to extend BreadthFirstIterator since some of the internals are likely to be irrelevant.

3) New interface for chordality inspection.  Hold off on this until (1) is merged.  The discussion will probably be around whether this should be one big interface including some optional methods (e.g. for returning a certificate, fill in edges, etc) vs breaking it down into subinterfaces.

@d-michail @jkinable does this sound right?

Toptachamann commented:
So, I need to ask what precisely I should do next:
1. Leave already reviewed implementation of the LexBFS in the ChordalityInspector, don't include MCS and then, after appropriate changes, substitute LexBFS with MCS in another pull request, which will also include changes to the way LexBFS in used now.
1. Add MCS to the ChordalityInspector and create, for example, two methods for checking chordality: one with LexBFS and one with MCS.
1. Remove LexBFS from ChordalityInspector with BucketList and add MCS instead. Add LexBFS with LexBFSIterator in another pull request.

d-michail commented:
@jsichi Yep, this sounds right. 

We could go for two vertex iterators, one `LexBFSIterator` and one `MCSIterator`, completely independent from the chordal graph recognition. 

Then one of the two could be used to implement the Chordality Inspector. In case some internal state is required for a more efficient implementation, we could add some additional public method in the iterators, to be able to query the state after execution.

jsichi commented:
Good idea, @d-michail.  So Timofey, does this sound right?  First create the two iterators (in the traverse package), and then add the ChordalityInspector on top (in the cycle package).  If MCS is faster, then use it by default, but if LexBFS is required for producing additional detail with the results, then use it automatically based on what the user requests.

Toptachamann commented:
I agree with this solution. It seems to me LexBFS doesn't produce any additional details except for the lexicographical order in contrast to the MCS order of MCS. That is why I'm going to do it the following way:
* After ChordalityInspector has processed the graph, it will contain order produced by MCSIterator (default option). This order will also be produced (lazily) after the call to getOrder().
* If the user calls getLexBFSOrder(), ChordalityInspector will produce this order lazily via ChordalityInspector.
* Chordality will also be recomputed via MCSIterator by default.

Toptachamann commented:
One note: in case of two orders (LexBFS and maximum cardinality) recomputations messes up the code a little bit. For example, if both orders have been computed already and the graph has been change, then after chordality recomputation maximum cardinality order is updated, but the LexBFS is not. This can be fixed by adding appropriate method, but then it adds complication for the user. Moreover, since recomputation is not consistent with other code, it gets even more confusing. 
So I decided not to add this feature.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,211 @@
+package org.jgrapht.alg.cycle;
```
Header AWOL

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,211 @@
+import java.util.*;
+
+/**
+ * Implementation of the lexicographical breadth-first search algorithm for chordal graph recognition.
```
This line needs to be updated now that MCS is in the picture too.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    /**
+     * Checks whether the inspected graph is chordal. Note: Every time this method is invoked, LexBFS order is recomputed.
```
Comment needs update.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/traverse/LexicographicalBfsIterator.java:
```diff
@@ -0,0 +1,329 @@
+ * @author Timofey Chudakov
+ * @since 1.8
+ */
+public class LexicographicalBfsIterator<V, E> extends AbstractGraphIterator<V, E> {
```
You're extending AbstractGraphIterator, but you're not actually implementing any of the GraphIterator interface.  So you should either make addTraversalListener throw unsupported, or else just extend Iterator instead of GraphIterator.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/traverse/MaximumCardinalityIterator.java:
```diff
@@ -0,0 +1,153 @@
+ * @param <E> the graph edge type.
+ * @author Timofey Chudakov
+ */
+public class MaximumCardinalityIterator<V, E> extends AbstractGraphIterator<V, E> {
```
Same comment w.r.t. extending AbstractGraphIterator.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/traverse/LexicographicalBfsIterator.java:
```diff
@@ -0,0 +1,329 @@
+ * @param <V> the graph vertex type.
+ * @param <E> the graph edge type.
+ * @author Timofey Chudakov
+ * @since 1.8
```
What is 1.8?

Toptachamann replied:
It is the Java version. I'll put the date there

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/traverse/LexicographicalBfsIterator.java:
```diff
@@ -0,0 +1,329 @@
+ * @author Timofey Chudakov
+ * @since 1.8
+ */
+public class LexicographicalBfsIterator<V, E> extends AbstractGraphIterator<V, E> {
```
Indeed, you need to implement `Iterator<V>` or actually implement the additional functionality of the `AbstractGraphIterator` (preferably by calling the listeners).

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/traverse/LexicographicalBfsIterator.java:
```diff
@@ -0,0 +1,329 @@
+
+    /**
+     * Returns a vertex with the lexicographically largest label, breaking ties arbitrarily.
+     * Recomputes cardinalities of its unvisited neighbours.
```
Remove the implementation detail from the Javadoc comment. The first sentence is enough to explain what it does. No need to state how it actually does it.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/traverse/LexicographicalBfsIterator.java:
```diff
@@ -0,0 +1,329 @@
+     * order. Labels aren't used explicitly, which results in time and space optimization.
+     *
+     * @author Timofey Chudakov
+     * @since 1.8
```
Same as above, what is 1.8?

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,211 @@
+ * A graph is chordal iff its the vertices can be arranged into a perfect elimination order. Perfect elimination
+ * order isn't unique. Both maximum cardinality search and lexicographical breadth-first search produces this order.
+ * As a result, chordality inspection can be performed in two way. Here it is done by default via MCS,
+ * because it outperforms LexBFS by a constant factor in practice. Nevertheless, lexicographical breadth-first
```
I would refrain from stating such facts without the proper reference. It there some particular paper that has done the experimental comparison and state that? Have you done some experiments and published the results somewhere?

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,211 @@
+     *
+     * @param graph the graph for which a chordality inspector to be created.
+     */
+    public ChordalityInspector(Graph<V, E> graph) {
```
I would prefer the class to have a single method 'isChordal' and people can change the behavior by the constructor. This means that you should be able to adjust the behavior of the class to either use LexBFS or MCS.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,211 @@
+     *
+     * @return the vertices of the {@code graph} in the lexicographical breadth-first order.
+     */
+    public List<V> getLexicographicalBfsOrder() {
```
This method is not needed anymore, the user can use directly the Iterator.

Toptachamann replied:
I have a question. There is no need to save computed order at all or ChordalityInspector can save computed ordering in order to, for example, return it as a perfect elimination order after the graph turned out to be chordal?

d-michail replied:
Better to cache the computed order and return it as a perfect elimination order.

Toptachamann replied:
Thanks

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,211 @@
+     *
+     * @return the vertices of the {@code graph} in the maximum cardinality order.
+     */
+    public List<V> getMaximumCardinalityOrder() {
```
Same as above.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,219 @@
+    /**
+     * Specifies {@link MaximumCardinalityIterator} as an internal iterator for this inspector.
+     */
+    public static final int MCS = 0;
```
Add an enumeration as an inner class here, called something like IterationOrder.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,219 @@
+    /**
+     * Iterator used for producing perfect elimination order.
+     */
+    private AbstractGraphIterator<V, E> orderIterator;
```
Change the type to either GraphIterator or Iterator which are interfaces.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,219 @@
+ *
+ * @param <V> the graph vertex type.
+ * @param <E> the graph edge type.
+ * @author March 2018
```
Typo

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/traverse/MaximumCardinalityIterator.java:
```diff
@@ -0,0 +1,223 @@
+ *
+ * @param <V> the graph vertex type.
+ * @param <E> the graph edge type.
+ * @author March 2018
```
typo

Toptachamann commented:
Thanks a lot

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    /**
+     * Creates a chordality inspector for {@code graph}, which uses as iterator defined by the second
```
typo:  "as iterator defined" -> "an iterator defined"

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,217 @@
+     * Creates a chordality inspector for {@code graph}, which uses as iterator defined by the second
+     * parameter as an internal iterator.
+     *
+     * @param graph          the graph for which a chordality inspector to be created.
```
"is to be created"

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    /**
+     * Returns the computed vertex order. In the case the inspected graph chordal, returned order
```
in the case where

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,217 @@
+     *
+     * @return computed vertex order.
+     */
+    public List<V> getOrder() {
```
Is there a more specific name we can give this, e.g. getEliminationOrder?

Toptachamann replied:
It is tricky, there is only the notion of "perfect elimination order", but I haven't encountered the notion of "elimination order". Either we can use the name "elimination order" and explain details in the Javadoc, or we can use "getComputedOrder".

jsichi replied:
getSearchOrder?

Toptachamann replied:
Great

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,217 @@
+ * four or more vertices have a chord, which is an edge that is not part of the cycle but connects two vertices
+ * of the cycle.
+ * <p>
+ * A graph is chordal iff its the vertices can be arranged into a perfect elimination order. Perfect elimination
```
I think you mean "More than one perfect elimination order may exist for a given graph.  Either maximum cardinality search or lexicographical breadth-first search can be used to produce such an order."

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/traverse/MaximumCardinalityIterator.java:
```diff
@@ -0,0 +1,224 @@
+ * For this iterator to work correctly the graph must not be modified during iteration. Currently
+ * there are no means to ensure that, nor to fail-fast. The results of such modifications are
+ * undefined.
+ *
```
Should note that only vertex events are fired by this iterator.

jsichi commented on jgrapht-core/src/test/java/org/jgrapht/traverse/LexBreadthFirstIteratorTest.java:
```diff
@@ -0,0 +1,175 @@
+ *
+ * @author Timofey Chudakov
+ */
+public class LexBreadthFirstIteratorTest {
```
Need test for events.

Toptachamann replied:
I implemented edgeTraversed firing. Do I understand it right: an iterator is cross-component if it iterates vertices in one connected component and then begins iterating vertices from another connected component? For example, BreadthFirstIterator isn't cross-component if it is created with vertices it should iterate first, otherwise, it iterates vertices across connected components. Both lexicographical breadth-first search and maximum cardinality search don't have any sense if they iterate vertices not across connected components. That is why they don't have to have an option to specify start vertices. Or I don't get it right and it has other meaning?

jsichi replied:
Yes, your understanding is correct.

Regarding the edge traversals, for the existing iterators such as DFS/BFS, there's a well-defined notion of which edge the traversal used.  Is this actually the case for MCS/LexBFS?  If it is, then you should only be firing for a specific edge, not all outgoing.  If it isn't, then you shouldn't be firing for edges at all (see TopologicalIterator).

Toptachamann replied:
Resting upon articles I've read there is no such definition. It is possible to define which edge is traversed on the analogy of BFS, but I think it'll be better not to do it. The reason is there are actually two possible definitions: all outgoing edges of a vertex or only those edges, which are incident to the vertices, which haven't been visited.

jsichi commented on jgrapht-core/src/test/java/org/jgrapht/alg/cycle/ChordalityInspectorTest.java:
```diff
@@ -0,0 +1,385 @@
+ *
+ * @author Timofey Chudakov
+ */
+public class ChordalityInspectorTest {
```
Need coverage for choosing MCS vs LexBFS.  Where relevant, parameterize the test and verify both ways.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,232 @@
+        if (graph.getType().isDirected()) {
+            this.graph = new AsUndirectedGraph<>(graph);
+        }
+        if (iterationOrder == IterationOrder.MCS) {
```
You should store the iteration order in a private variable and use it instead of doing instanceof.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,232 @@
+     * @return the type of iterator used in this {@code ChordalityInspector}
+     */
+    public IterationOrder getIterationOrder() {
+        if (orderIterator instanceof LexBreadthFirstIterator) {
```
Use a private field which takes its value in the constructor.

d-michail commented:
One last change, to store the iteration order.

d-michail commented:
Looks good!

Toptachamann commented:
Thank you. This means a lot for me.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java:
```diff
@@ -0,0 +1,233 @@
+ * Both lexicographical BFS and maximum cardinality search run in $\mathcal{O}(|V| + |E|)$. Checking whether given order
+ * is the perfect elimination order via {@link ChordalityInspector#isPerfectEliminationOrder(List)} takes
+ * $\mathcal{O}(|V| + |E|)$ as well. So, overall time complexity of the method
+ * {@link ChordalityInspector#isChordal()} is $\mathcal{O}(|V| + |E|)$.
```
Would it make sense to add a line of text stating why you would use BFS instead of LEX_BFS (or vice versa)?

Toptachamann replied:
Actually, it was added (the note was "the maximum cardinality search is used as a default iterator because it performs better in practice" or something similar). This statement is based only on my personal tests (random instances and complete graphs were used). MCS performs 1.5 times better than LexBFS on average. As it is not the official result and the algorithms weren't tested on all types of graphs, @d-michail suggested removing this note.

It seems to me, the note could sound something like this "Both algorithms have linear time complexity. The default option for ChordalityInspector is MaximumCardinalityIterator. This can be changed with an appropriate constructor parameter".

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/traverse/LexBreadthFirstIterator.java:
```diff
@@ -0,0 +1,380 @@
+ * Iterator chooses vertex with lexicographically largest label and returns it. It breaks ties
+ * arbitrarily. For more information on lexicographical BFS see the following papers: <a
+ * href="https://pdfs.semanticscholar.org/d4b5/a492f781f23a30773841ec79c46d2ec2eb9c.pdf">
+ * <i>Lexicographic Breadth First Search – A Survey</i></a> and <a
```
Please include the author in the reference. If the link goes dead, it might be hard to trace the paper back.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/traverse/MaximumCardinalityIterator.java:
```diff
@@ -0,0 +1,226 @@
+ * For every vertex in graph its cardinality is defined as the number of its neighbours, which
+ * have been already visited by this iterator. Iterator chooses vertex with the maximum cardinality,
+ * breaking ties arbitrarily. For more information of maximum cardinality search see
+ * <a href="http://www.ii.uib.no/~pinar/MCS-M.pdf"><i>Maximum Cardinality Search for Computing Minimal Triangulations</i></a>.
```
Same here: please use proper citation

jkinable commented:
I left a few comments on the documentation. Should be easy to fix. Nice job. One more request:

Can you add a 'isChordal(Graph<.> g)' method to GraphTests?

Toptachamann commented:
Thanks.
Besides MCS and LexBFS I've encountered only straightforward algorithm, which is much slower. 
In my opinion, there are three possible directions to develop in this area further:
* Implement methods for finding the maximum clique and the chromatic number of a chordal graph. This can be done with the perfect elimination order produced by ChordalityInspector.
* Use MCS and LexBFS in other applications. For example, a modification of MCS can compute minimal triangulation. This is described in this [article](http://www.ii.uib.no/~pinar/MCS-M.pdf). I went only briefly through this article and can't say for sure whether the algorithm is efficient and whether it hasn't been implemented already in jgrapht. But the idea is to use these algorithms for their applications.
* Develop WeakChordalityInspector (I'm currently working in this direction). There exist two efficient algorithms for this problem described [here](http://www.ii.uib.no/~skeide/rec_wcg.pdf) and [here](https://webdocs.cs.ualberta.ca/~hayward/papers/iawcg.pdf). The algorithms are a little bit more sophisticated and have own trade-offs. Both have O(m^2) running time complexity but different memory complexity - O(m^2) and O(n + m).

jkinable commented:
For weakly chordal graphs, also check these references: http://www.cs.uoi.gr/~stavros/C-Papers/C-2004-SODA.pdf and http://doc.sagemath.org/html/en/reference/graphs/sage/graphs/weakly_chordal.html

One more request for the current PR. If your chordality inspector decides that a graph is *not* chordal, it means that the graph contains at least one hole (http://mathworld.wolfram.com/GraphHole.html, also known as chordless cycle). The existence of a hole *certifies* that the graph indeed is not chordless. Can you add a method that provides such a certificate? 

Perhaps:
/** If the graph is not chordal, by definition this implies that the graph contains at least one hole (http://mathworld.wolfram.com/GraphHole.html, also known as chordless cycle). To certify that the graph is not chordal, this method returns one such hole. **/
GraphPath<> getHole();

Toptachamann commented:
This is a good idea. Now I have to focus a little bit on my proposal and after that, I'll add it.

jsichi commented:
@jkinable would it be OK if we merged this one now and then add the hole certificates in a followup?  We previously discussed a followup for adding a common ChordalityAlgorithm interface over this and the existing CliqueMinimalSeparatorDecomposition, so it might make sense to do those together since it also produces a certificate in the form of the fill edges.

(The history on this PR is quite long already so it would be nice to work from a fresh baseline.)

jkinable commented:
Sure, I'm fine with that.

jsichi commented:
Thanks Timofey!  Merging now.

jkinable commented:
I've updated the TODO wiki to reflect the follow-up changes:
https://github.com/jgrapht/jgrapht/wiki/TODOs

Toptachamann commented:
Thank you for your reviews, they are very helpful.
It is a very cool feeling to make this contribution.
